pkg/role/datastore.go

```diff
@@ -0,0 +1,162 @@
+package role
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/portworx/kvdb"
+)
+
+// Datastore provides APIs for storing and fetching SDK Roles
+// from a persistent datastore
+type Datastore interface {
+
+	// Exists returns true if the role exists in the datastore
+	Exists(roleName string) (bool, error)
+
+	// Update updates the role in the datastore. It will return an error
+	// if the role does not exist in the datastore
+	Update(roleName string, sdkRole *api.SdkRole) error
+
+	// Put puts the role in the datastore. It will overwrite the
+	// role object if it already exists
+	Put(roleName string, sdkRole *api.SdkRole) error
+
+	// Get gets the role object from the datastore for the given name
+	Get(roleName string) (*api.SdkRole, error)
+
+	// Create creates the role in the datastore. It will return an error
+	// if the role with the name already exists
+	Create(roleName string, sdkRole *api.SdkRole) error
+
+	// EnumerateRoleNames enumerates all the roles from the datastore and returns their names
+	EnumerateRoleNames() ([]string, error)
+
+	// Delete deletes the role from the datastore
+	Delete(roleName string) error
+}
+
+var (
+	// ErrRoleNotFound is returned when role is not found in datastore
+	ErrRoleNotFound = errors.New("role not found in datastore")
+	// ErrRoleExists is returned when a role with the same name exists
+	ErrRoleExists = errors.New("role exists with the same name")
+)
+
+const (
+	rolePrefix = "cluster/roles"
+)
+
+// Simple function which creates key for Kvdb
+func prefixWithName(name string) string {
+	return rolePrefix + "/" + name
+}
+
+type kvRoleDatastore struct {
+}
+
+// NewKvdbRoleDatastore returns a kvdb implementation of Datastore interface
+func NewKvdbRoleDatastore() (Datastore, error) {
```

Comment:
This is a style change, but instead of returning an interface, return the object. Let the caller decide if they want to use the interface. It also helps with unit tests.

If you really want to return an interface, then have `New..` call `new...` which returns the object, that way the unit tests can have access to the real object.

Reply:
Makes sense. Will do the change.